guides/errors/execution_errors.md

```diff
@@ -0,0 +1,80 @@
+---
+layout: guide
+search: true
+section: Errors
+title: Top-level "errors"
+desc: The top-level "errors" array and how to use it.
+index: 1
+---
+
+The GraphQL specification [allows for a top-level `"errors"` key](http://facebook.github.io/graphql/October2016/#sec-Errors) in the response which may contain information about what went wrong during execution. For example:
+
+```ruby
+{
+  "errors" => [ ... ]
+}
+```
+
+The response may include _both_ `"data"` and `"errors"` in the case of a partial success:
+
+```ruby
+{
+  "data" => { ... } # parts of the query that ran successfully
+  "errors" => [ ... ] # errors that prevented some parts of the query from running
+}
+```
+
+## When to Use Top-Level Errors
+
+In general, top-level errors should only be used for exceptional circumstances when a developer should be made aware that the system had some kind of problem.
+
+For example, the GraphQL specification says that when a non-null field returns `nil`, an error should be added to the `"errors"` key. This kind of error is not recoverable by the client. Instead, something on the server should be fixed to handle this case.
+
+When you want to notify a client some kind of recoverable issue, consider making error messages part of the schema, for example, as in {% internal_link "mutation errors", "/mutations/mutation_errors" %}.
+
+## Adding Errors to the Array
+
+In GraphQL-Ruby, you can add entries to this array by raising `GraphQL::ExecutionError` (or a subclass of it), for example:
+
+```ruby
+raise GraphQL::ExecutionError, "Can't continue with this query"
+```
+
+When this error is raised, its `message` will be added to the `"errors"` key and GraphQL-Ruby will automatically add the `line`, `column` and `path` to it. So, the above error might be:
+
+```ruby
+{
+  "errors" => [
+    {
+      "message" => "Can't continue with this query",
+      "locations" => [
+        {
+          "line" => 2,
+          "column" => 10,
+        }
+      ],
+      "path" => ["user", "login"],
+    }
+  ]
+}
+```
+
+## Customizing Error JSON
+
+The default error JSON includes `"message"`, `"locations"` and `"path"`. You can customize this in two ways:
+
+- Pass `options:` when raising an error, for example:
+  ```ruby
+  raise GraphQL::ExecutionError.new("Something went wrong", options: { "code" => "BROKEN" })
+  ```
+  In this case, `"code" => "BROKEN"` will be added to the error JSON.
```

Comment:
I'm pretty sure the spec says any custom properties should be nested under `extensions`. Right now `options` just  get's merged into the error itself. I was planning on doing a PR to fix/support this anyway, but it should probably be mentioned or accounted for in the docs as well.

See section below http://facebook.github.io/graphql/draft/#example-08b62

Reply:
> Previous versions of this spec did not describe the extensions entry for error formatting. 
 While non‐specified entries are not violations, they are still discouraged.

I definitely had not seen that before.

Reply:
Thanks for the pointer, I updated it a bit in 81b22820 to make that suggestion and use `"extensions"` in the examples. Maybe a nice feature would be to add an `extensions:` keyword that merged that value into any previously-added extensions.